ktor-server/ktor-server-core/jvm/src/io/ktor/response/ApplicationResponseFunctions.kt

```diff
@@ -12,22 +12,29 @@ import io.ktor.http.content.*
 import io.ktor.util.*
 import io.ktor.utils.io.*
 import java.io.*
+import kotlin.reflect.*
 
 /**
  * Sends a [message] as a response
  */
-@Suppress("NOTHING_TO_INLINE")
-public suspend inline fun ApplicationCall.respond(message: Any) {
-    response.pipeline.execute(this, message)
+@OptIn(ExperimentalStdlibApi::class)
+public suspend inline fun <reified T> ApplicationCall.respond(message: T) {
+    response.pipeline.intercept(ApplicationSendPipeline.Before) {
+        call.attributes.put(ResponseTypeAttributeKey, typeOf<T>())
```

Comment:
The problem with such an approach is that if a response object gets replaced, the paired type keeps remain that may lead to strange consequences. However, this is a good starting point. Why can't we put this type immediately instead of interception?

Reply:
sure, I don't know why I made the interceptor here. but again, it's still a draft PR that will be updated